spec/module_validator.rb

```diff
@@ -0,0 +1,167 @@
+    unless value.is_a?(Array)
+      record.errors.add(attribute, "#{attribute} must be an array")
+      return
+    end
+
+    if attribute == :references
+      reference_keys = []
+      value.each { |k, _v| reference_keys << k }
+      value = reference_keys
+    end
+
+    invalid_options = value - options[:in]
+    message = "contains invalid values #{invalid_options.inspect} - only #{options[:in].inspect} is allowed"
+
+    if invalid_options.any?
+      record.errors.add(attribute, :array_inclusion, message: message, value: value)
+    end
+  end
+end
+
+class ModuleValidator < SimpleDelegator
+  include ActiveModel::Validations
+
+  validate :validate_filename_is_snake_case, :validate_reference_ctx_id, :validate_author_bad_chars, :validate_target_platforms
+
+  attr_reader :mod
+
+  def initialize(mod)
+    super
+    @mod = mod
+  end
+
+  #
+  # Acceptable Stability ratings
+  #
+  VALID_STABILITY_VALUES = [
+    Msf::CRASH_SAFE,
+    Msf::CRASH_SERVICE_RESTARTS,
+    Msf::CRASH_SERVICE_DOWN,
+    Msf::CRASH_OS_RESTARTS,
+    Msf::CRASH_OS_DOWN,
+    Msf::SERVICE_RESOURCE_LOSS,
+    Msf::OS_RESOURCE_LOSS
+  ]
+
+  #
+  # Acceptable Side-effect ratings
+  #
+  VALID_SIDE_EFFECT_VALUES = [
+    Msf::ARTIFACTS_ON_DISK,
+    Msf::CONFIG_CHANGES,
+    Msf::IOC_IN_LOGS,
+    Msf::ACCOUNT_LOCKOUTS,
+    Msf::SCREEN_EFFECTS,
+    Msf::AUDIO_EFFECTS,
+    Msf::PHYSICAL_EFFECTS
+  ]
+
+  #
+  # Acceptable Reliability ratings
+  #
+  VALID_RELIABILITY_VALUES = [
+    Msf::FIRST_ATTEMPT_FAIL,
+    Msf::REPEATABLE_SESSION,
+    Msf::UNRELIABLE_SESSION
+  ]
+
+  #
+  # Acceptable site references
+  #
+  VALID_REFERENCE_CTX_ID_VALUES = %w[CVE CWE BID MSB EDB US-CERT-VU ZDI URL WPVDB PACKETSTORM LOGO SOUNDTRACK OSVDB VTS OVE]
+
+  def validate_notes_values_are_arrays
+    notes.each do |k, v|
+      unless v.is_a?(Array)
+        errors.add :notes, "#{k}'s value must be an array, got #{v.inspect}"
+      end
+    end
+  end
+
+  def validate_crash_safe_not_present_in_stability_notes
+    if rank == Msf::ExcellentRanking && !stability.include?(Msf::CRASH_SAFE)
+      errors.add :stability, "must have CRASH_SAFE value if module has an ExcellentRanking, instead found #{stability.inspect}"
+    end
+  end
+
+  def validate_filename_is_snake_case
+    unless file_path.split('/').last.match?(/^[a-z0-9]+(?:_[a-z0-9]+)*\.rb$/)
+      errors.add :file_path, "must be snake case, instead found #{file_path.inspect}"
+    end
+  end
+
+  def validate_reference_ctx_id
+    references_ctx_id_list = references.map(&:ctx_id)
+    invalid_references = references_ctx_id_list - VALID_REFERENCE_CTX_ID_VALUES
+
+    invalid_references.each do |ref|
+      if ref == 'NOCVE'
+        errors.add :references, "#{ref} please include NOCVE values in the 'notes' section, rather than in 'references'."
+      elsif ref == 'AKA'
+        errors.add :references, "#{ref} please include AKA values in the 'notes' section, rather than in 'references'."
```

Comment:
Not a blocker: We shouldn't use full stops here, as we'd normally get back an array of errors back - that the user could decide to join together with `.`, or use in an HTML page separted by `<li>...</li>` - depending on how they want to present the data, the extra `.` at the end of the string might break things

```suggestion
        errors.add :references, "#{ref} please include AKA values in the 'notes' section, rather than in 'references'"
```